### PR TITLE
add new pytest_handlecrashitem hook

### LINKED ISSUE #650: pytest-rerunfailures for crashed nodes
There's a hook for crashed nodes, but it doesn't let you clean-up the test failure. I'm looking for a way to contain a crashed test and rerun it.

### PR DESCRIPTION
**Pull request by xoviat**
fixes #650, allows a hook to handle a crashitem (if handled, will be put back into test queue).

hook runs in controller, is responsible for communicating status to nodes if needed.

note: I thought of using a file for communication, but ideas welcome. rerunfailures uses `pytest_runtest_protocol`, so I will need to pass a message from this hook (in the controller) to that hook (in the node).

linked pr: pytest-dev/pytest-rerunfailures#158

### CODE REVIEW

**Comment by xoviat** [comment]
@RonnyPfannschmidt thoughts?

File: src/xdist/scheduler/load.py
Comment (RonnyPfannschmidt) [inline src/xdist/scheduler/load.py:206]: Always pop, the hook can put it back in

This hook may nerd company to manage detail like max worker restart

Im still not sure if its a good idea to just restart those crashers, they are tests that anhillated a python process, so it's the kind of flaky ness that requires different attention

Comment (xoviat) [reply src/xdist/scheduler/load.py:206]: One reason for the crash is that a timeout occurred. In my case, I combine pytest-timeout with rerunfailures to rerun tests that timed out. However, pytest-rerunfailures is unable to catch this because it runs inside the worker.

**Comment by xoviat** [comment]
Is this better?

File: src/xdist/scheduler/load.py
Comment (xoviat) [reply src/xdist/scheduler/load.py]: What do you propose? AFAIK, this has to crash the worker, because there's no other way to stop running python code other than to kill it.

Comment (RonnyPfannschmidt) [reply src/xdist/scheduler/load.py]: the timeout plugin module has multiple ways to enable this, but i'll have to read into the details

**Comment by xoviat** [comment]
Actually, I may have a solution to this problem.

**Comment by xoviat** [comment]
Even with the timeout plugin patched, there is no other way in certain circumstances to stop the test other than to kill the worker. This is the conclusion after applying this patch: https://github.com/xoviat/pytest-timeout/commit/d84f0608ab3b216ca3413830db29823ae700f005

Only after multiple attempts have been made to stop the test is the worker killed. And it is killed: https://github.com/pyinstaller/pyinstaller/runs/2420166528#step:10:1124

Meaning: the attempts to stop the test failed, and we must kill the worker (on windows) and rerun the test.

**Comment by xoviat** [comment]
@nicoddemus Ideas?

**Comment by xoviat** [comment]
The hook needs to indicate whether the test should be rerun, and optionally, needs to generate a failure report. Perhaps a better approach is to call the hook from handle_crashitem with the report, crashitem, and scheduler as parameters, and then add a method to the scheduler that would allow the hook to add the crashitem back.

**Comment by nicoddemus** [comment]
> @nicoddemus Ideas?

Yeah AFAIK that's the only way to stop the test from running.

About the PR: I'm definitely OK with the design as is: trivial on xdist's side and gives power to plugins/users on how to handle crash items. 

@RonnyPfannschmidt do you have any objections?

**Comment by RonnyPfannschmidt** [comment]
No, we may want to declare that hook experimental

**Comment by xoviat** [comment]
Anything else?

**Comment by nicoddemus** [comment]
Yes, if you please:

- [ ] Document the new hook, preferably with a simple example, like marking the item as crashed (adding a message to the report?) and re-scheduling it.
- [ ] A simple integration test that ensures the new hook is being called (we don't want to regress it).
- [ ] A new `improvement` CHANGELOG entry 

Feel free to ask for guidance in any of the above. 👍 

Thanks!

**Comment by xoviat** [comment]
Done.

**Comment by xoviat** [comment]
@nicoddemus Is there anything else?

File: CHANGELOG.rst
Comment (nicoddemus) [inline CHANGELOG.rst:9]: Sorry for not being clearer, you should actually create a file `changelog/650.feature.rst`, just with this contents:

```rst
Added new ``pytest_handlecrashitem`` hook to allow handling and rescheduling crashed items.
```

We use [towncrier](https://pypi.org/project/towncrier/) to manage the changelog, which is automatically generated before each release.

File: src/xdist/newhooks.py
Comment (nicoddemus) [inline src/xdist/newhooks.py:72]: Nice docs!

**Review by nicoddemus — COMMENTED** [review]
Thanks @xoviat, sorry for the delay!

File: CHANGELOG.rst
Comment (nicoddemus) [reply CHANGELOG.rst]: Went ahead and did it myself. 👍

**Comment by nicoddemus** [comment]
Also rebased in the latest `master` to get a green CI. 👍

**Review by nicoddemus — APPROVED** [review]
Thanks @xoviat for the feature and patience to carry this on!

Let's wait for @RonnyPfannschmidt to do another round of review before merging this. 👍

**Review by RonnyPfannschmidt — APPROVED** [review]
Looks good to me, im not able to do a deep review but I'm under the impression I won't have to do one

**Pull request merged**
